doc/developer/design/20220411_reclocking_implementation.md

```diff
@@ -0,0 +1,299 @@
+  ((2, +4),  1649686079487,   +1)    # (4)
+```
+
+Entry (1) indicates that offsets 0-42 (inclusive) in partition 1 are assigned a
+Materialize timestamp of Monday, April 11, 2022 14:07:56.392 PM UTC. Entry (2)
+indicates that offsets 0-40 in partition 2 are assigned the same timestamp.
+Entry (3) indicates that offsets 43-45 in partition 1 are assigned a Materialize
+timestamp of 14:07:59.487 UTC on the same day. Entry (4) indicates that offsets
+41-44 in partition 2 are assigned the same timestamp.
+
+Notice how the offsets are described by their *increments*, not their absolute
+value. This allows the collection to be compacted using `partition` as the
+compaction key. This generalizes to the other reclockable source types.
+
+#### `remap` policy
+
+The mapping must always preserve the order of events from the upstream source
+whenever such an order is present. If transactional boundaries are present in
+the upstream source, events in the same transaction must be mapped to the
+same timestamp.
+
+There is explicitly no commitment to the granularity of the mapping. Any number
+of upstream events may be mapped to the same Materialize timestamp, provided the
+above properties are upheld.
+
+Concretely:
+
+  * The mapping for a file source must maintain the line order of the file.
+
+  * The mapping for a Kafka source depends on whether Debezium's transactional
+    metadata topic is in use:
+
+      * If Debezium's transactional metadata topic is not in use, the mapping
+        must maintain the offset order of the Kafka topic within each partition.
+        It does not need to maintain order across partitions.
+
+      * If Debezium's transactional metadata topic is in use, the mapping must
+        maintain the order of transactions as determined by the transactional
+        metadata topic. TODO: can we be more specific here?
+
+  * The mapping for a Kinesis source must maintain order by sequence number
+    within each shard. It does not need to maintain order across shards.
+
+  * The mapping for a PostgreSQL source must maintain order of the commit LSN
+    for each transaction. It must present all changes in a single transaction at
+    the same timestamp. TODO: what is the difference between "end LSN" and
+    "commit LSN" in the PostgreSQL logical replication protocol?
+
+  * The mapping for an S3 source must maintain the line order within each
+    object. It does not need to maintain order across objects.
+
+#### Mechanics
+
+For reclockable sources, the ingestion pipeline will be separated into
+two phases:
+
+1. The production of the `remap` collection.
+
+2. The ingestion of the source data according to the `remap` collection.
+
+Separating these components is important because phase two is embarrassingly
+parallel and can be made highly available without consensus, whereas phase one
+is inherently unparallelizable and requires consensus to be made highly
+available. The separation also sets us up to support bring-your-own (BYO)
```

Comment:
it seemed an implication of one of your comments below re: sinks is that this separation makes re-ingesting sinks doable?

Reply:
Frank has convinced me that that comment is wrong and that re-ingesting sinks is orthogonal to how we restructure the source ingestion pipeline.